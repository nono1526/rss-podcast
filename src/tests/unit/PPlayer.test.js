import PPLayer from '@src/components/PPLayer.vue'
import { shallowMount } from '@vue/test-utils'

describe('PPLayer.vue', () => {
  let mountFunction
  const fakeAudio = new Audio()

  beforeEach(() => {
    mountFunction = options => shallowMount(PPLayer, {
      ...options,
      propsData: {
        audioElement: fakeAudio
      }
    })
  })
  it('match snapshot', () => {
    const wrapper = mountFunction({})
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should play when props.url change', async () => {
    const play = jest.fn()
    fakeAudio.play = play
    const wrapper = mountFunction({})
    await wrapper.setProps({ url: 'a. mp4' })
    expect(play).toBeCalled()
  })

  it('should trigger @end when audio ended', () => {
    const wrapper = mountFunction({})
    fakeAudio.dispatchEvent(new Event('ended'))
    expect(wrapper.emitted('ended')).toHaveLength(1)
  })

  it('should click nextBtn work', () => {
    const wrapper = mountFunction({})
    const nextBtn = wrapper.findComponent('[data-test="next-btn"]')
    nextBtn.vm.$emit('click')
    expect(wrapper.emitted('click:next')).toHaveLength(1)
  })

  it('should click prev work', () => {
    const wrapper = mountFunction({})
    const prevBtn = wrapper.findComponent('[data-test="prev-btn"]')
    prevBtn.vm.$emit('click')
    expect(wrapper.emitted('click:prev')).toHaveLength(1)
  })

  it('click play btn toggle isPlaying ', async () => {
    const wrapper = mountFunction({})
    const playBtn = wrapper.findComponent('[data-test="play-btn"]')
    playBtn.vm.$emit('click')
    expect(wrapper.emitted('update:isPlaying').pop()).toEqual([true])
    await wrapper.setProps({
      isPlaying: true
    })
    playBtn.vm.$emit('click')
    expect(wrapper.emitted('update:isPlaying').pop()).toEqual([false])
  })

  it('should isPlaying state control play() and pause()', async () => {
    const play = jest.fn()
    const pause = jest.fn()
    fakeAudio.play = play
    fakeAudio.pause = pause
    const wrapper = mountFunction({})
    await wrapper.setProps({
      isPlaying: true
    })
    expect(play).toBeCalled()
    await wrapper.setProps({
      isPlaying: false
    })
    expect(pause).toBeCalled()
  })

  it('should sync currentTime', async () => {
    const wrapper = mountFunction({})
    fakeAudio.dispatchEvent(new Event('timeupdate'))
    expect(wrapper.emitted('update:currentTime')).toHaveLength(1)
  })

  it('playedTime correctly', async () => {
    const wrapper = mountFunction({})
    expect(wrapper.vm.playedTime).toBe('00:00')
    await wrapper.setProps({ currentTime: 10 })
    expect(wrapper.vm.playedTime).toBe('00:10')
    await wrapper.setProps({ currentTime: 70 })
    expect(wrapper.vm.playedTime).toBe('01:10')
    await wrapper.setProps({ currentTime: 60 * 60 })
    expect(wrapper.vm.playedTime).toBe('01:00:00')
  })

  it('restTime correctly', async () => {
    const wrapper = mountFunction({})
    wrapper.vm.duration = 10
    expect(wrapper.vm.restTime).toBe('00:10')
    wrapper.vm.duration = 60 * 60
    expect(wrapper.vm.restTime).toBe('01:00:00')
    await wrapper.setProps({ currentTime: 60 })
    expect(wrapper.vm.restTime).toBe('59:00')
    await wrapper.setProps({ currentTime: 61 })
    expect(wrapper.vm.restTime).toBe('58:59')
  })

  it('set volume', async () => {
    const wrapper = mountFunction({})
    const volumeSlider = wrapper.findComponent('[data-test="volume-slider"]')
    volumeSlider.vm.$emit('input', 100)
    expect(fakeAudio.volume).toBe(1)
    volumeSlider.vm.$emit('input', 0)
    expect(fakeAudio.volume).toBe(0)
    volumeSlider.vm.$emit('input', 50)
    expect(fakeAudio.volume).toBe(0.5)
  })
})
