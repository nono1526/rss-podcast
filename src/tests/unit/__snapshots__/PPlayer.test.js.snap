// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PPLayer.vue match snapshot 1`] = `
<div
  class="bottom-0 w-full bg-blue-50 border-t h-20 flex hidden"
>
  <img
    class="h-full cursor-pointer"
    src=""
  />
  <div
    class="flex flex-col h-full justify-center mx-5 text-sm w-48 cursor-pointer"
  >
    <div
      class="truncate"
    />
    <div
      class="text-xs text-gray-400 truncate"
    />
  </div>
  <div
    class="flex-1 flex flex-col justify-center"
  >
    <div
      class="w-8/12 mx-auto"
    >
      <div
        class="flex"
      >
        <p-btn-stub
          data-test="next-btn"
          disabled="true"
          icon=""
        />
        <p-btn-stub
          data-test="replay-10-btn"
          disabled="true"
          icon=""
        />
        <p-btn-stub
          class="mx-auto"
          data-test="play-btn"
          disabled="true"
          icon=""
        />
        <p-btn-stub
          data-test="quick-10-btn"
          disabled="true"
          icon=""
        />
        <p-btn-stub
          data-test="prev-btn"
          disabled="true"
          icon=""
        />
      </div>
      <div
        class="w-full flex items-center"
      >
        <span
          class="mr-3 text-gray-500 w-14 text-right text-sm"
        >
          00:00
        </span>
        <p-slider-stub
          max="0"
          value="0"
        />
        <span
          class="ml-3 w-14 text-gray-500 text-sm"
        >
          00:00
        </span>
      </div>
    </div>
  </div>
  <div
    class="w-36 flex items-center mr-5"
  >
    <svg
      class="mx-2"
      fill="#515151"
      height="32px"
      viewBox="0 0 24 24"
      width="32px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M14 8.83v6.34L11.83 13H9v-2h2.83L14 8.83M16 4l-5 5H7v6h4l5 5V4z"
      />
    </svg>
    <p-slider-stub
      data-test="volume-slider"
      max="0"
      value="0"
    />
    <svg
      class="mx-2"
      fill="#515151"
      height="32px"
      viewBox="0 0 24 24"
      width="32px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0 0h24v24H0V0z"
        fill="none"
      />
      <path
        d="M3 9v6h4l5 5V4L7 9H3zm7-.17v6.34L7.83 13H5v-2h2.83L10 8.83zM16.5 12c0-1.77-1.02-3.29-2.5-4.03v8.05c1.48-.73 2.5-2.25 2.5-4.02zM14 3.23v2.06c2.89.86 5 3.54 5 6.71s-2.11 5.85-5 6.71v2.06c4.01-.91 7-4.49 7-8.77 0-4.28-2.99-7.86-7-8.77z"
      />
    </svg>
  </div>
</div>
`;
